File: src/js/src/data/CabinOffering.ts
import { EventEmitter } from 'node:events';
import type { Price } from './types';
import type { Departure } from './Departure';
import type { CabinGrade } from './CabinGrade';
import type { Ship } from './Ship';
import type { ISwotaAvailabilityClient } from '../availability/ISwotaAvailabilityClient';

/**
 * Observable state of a {@link CabinOffering}'s live-availability fetch (see
 * {@link CabinOffering.getAvailableCabinsAsync}).
 *
 * - `'static'`: no live client wired (`'v1'`/`'v3'` formats). The static
 *   {@link CabinOffering.availableCabins} snapshot IS the final answer —
 *   there is, and never will be, a live fetch. Set at construction and never
 *   changes.
 * - `'not-fetched'`: a live client IS wired (`'swota'` format), but
 *   {@link CabinOffering.getAvailableCabinsAsync} hasn't been called yet. Set
 *   at construction for `'swota'` offerings.
 * - `'loading'`: {@link CabinOffering.getAvailableCabinsAsync} has been
 *   called and the live fetch is in flight.
 * - `'loaded'`: the live fetch resolved successfully.
 * - `'failed'`: the live fetch threw (after whatever retries the client
 *   itself performs — this state layer adds no retry behavior of its own,
 *   it only surfaces the terminal failure).
 */
export type CabinAvailabilityState = 'static' | 'not-fetched' | 'loading' | 'loaded' | 'failed';

/**
 * A cabin grade made available on a specific departure, with prices per
 * currency. This is the join between a Departure and a CabinGrade — the node
 * where pricing lives. Navigable to its departure, grade, and ship.
 */
export class CabinOffering {
  /** Cabin grade code (source-market "Category", e.g. "DS"). */
  readonly code: string;
  /** Human label (source-market "SuperCategory", e.g. "DARWIN SUITE"). */
  readonly name: string;
  /**
   * Static available-cabins snapshot. Untouched by the `'swota'` format: under
   * `'v1'` this is a real value from the flat file, under `'v3'`/`'swota'` it
   * is `maxOccupancy` used as a placeholder. Always prefer
   * {@link getAvailableCabinsAsync} over reading this directly — it resolves
   * live availability when a live client is wired (`'swota'` format) and
   * falls back to this static value otherwise.
   */
  readonly availableCabins: number | null;
  /**
   * The RAW, unstripped source `VoyageID` this offering belongs to (e.g.
   * `"_@FNALA04-260906"`), used as the `voyageId` argument to
   * {@link ISwotaAvailabilityClient.getAvailableCabins}. The live SWOTA REST
   * API requires this exact raw form — it is NOT the same as the stripped
   * departure code. Only set when a live client is wired (`'swota'` format);
   * `undefined` for `'v1'`/`'v3'`.
   */
  readonly voyageId: string | undefined;

  private readonly _prices = new Map<string, Price>();
  private _departure!: Departure;
  private _cabinGrade: CabinGrade | null = null;
  private readonly _liveClient: ISwotaAvailabilityClient | undefined;
  private _liveAvailability: Promise<number | null> | undefined;
  private readonly _events = new EventEmitter();
  private _availabilityState: CabinAvailabilityState;
  private _lastKnownAvailableCabins: number | null;

  constructor(
    code: string,
    name: string,
    availableCabins: number | null,
    voyageId?: string,
    liveClient?: ISwotaAvailabilityClient
  ) {
    this.code = code;
    this.name = name;
    this.availableCabins = availableCabins;
    this.voyageId = voyageId;
    this._liveClient = liveClient;
    if (this._liveClient) {
      this._availabilityState = 'not-fetched';
      this._lastKnownAvailableCabins = null;
    } else {
      this._availabilityState = 'static';
      this._lastKnownAvailableCabins = availableCabins;
    }
  }

  /** Current state of the live-availability fetch. See {@link CabinAvailabilityState}. */
  get availabilityState(): CabinAvailabilityState {
    return this._availabilityState;
  }

  /**
   * Sync, side-effect-free snapshot of whatever's currently known: `null`
   * while `'not-fetched'`/`'loading'`/`'failed'`, the resolved value once
   * `'loaded'`, or the static {@link availableCabins} value immediately when
   * {@link availabilityState} is `'static'`. Never triggers a fetch — safe to
   * poll/read repeatedly, e.g. from a render loop.
   */
  get lastKnownAvailableCabins(): number | null {
    return this._lastKnownAvailableCabins;
  }

  /**
   * Registers `listener` to be called whenever {@link availabilityState} (and
   * therefore {@link lastKnownAvailableCabins}) changes:
   * `'not-fetched'`→`'loading'`, then `'loading'`→`'loaded'`/`'failed'`.
   * Never fires for `'static'` offerings — that state never changes.
   *
   * @returns An unsubscribe function.
   */
  onAvailabilityChange(listener: (offering: CabinOffering) => void): () => void {
    const wrapped = (): void => listener(this);
    this._events.on('change', wrapped);
    return () => {
      this._events.off('change', wrapped);
    };
  }

  private _transition(state: CabinAvailabilityState, resolvedValue?: number | null): void {
    this._availabilityState = state;
    if (resolvedValue !== undefined) {
      this._lastKnownAvailableCabins = resolvedValue;
    }
    this._emitChange();
  }

  /**
   * Emits the `'change'` event, swallowing any exception a listener throws.
   *
   * `EventEmitter.emit()` re-throws synchronously if a listener throws (Node
   * has no default `'error'` handler installed here). Every call site of
   * this method runs from inside the `.then()`/`.catch()` chain that
   * resolves/rejects {@link _liveAvailability} (via {@link _transition}) or
   * from the synchronous `'loading'` kickoff in
   * {@link getAvailableCabinsAsync} — in both cases, letting a listener's
   * exception propagate out of this method would let it get caught by that
   * surrounding chain and mis-attributed as a fetch failure, corrupting an
   * otherwise-successful result. By the time this is called, state and
   * {@link _lastKnownAvailableCabins}/{@link _liveAvailability} are already
   * finalized for this transition, so a listener's exception has nothing
   * left to corrupt — it's safe to just swallow it here. Mirrors .NET's
   * `CabinOffering.RaiseAvailabilityChanged`.
   */
  private _emitChange(): void {
    try {
      this._events.emit('change', this);
    } catch {
      // Deliberately swallowed -- see doc comment above.
    }
  }

  get departure(): Departure {
    return this._departure;
  }

  /** The cabin grade reference (null if the category is absent from cabingrades). */
  get cabinGrade(): CabinGrade | null {
    return this._cabinGrade;
  }

  /** The ship this offering sails on, via its departure. */
  get ship(): Ship | null {
    return this._departure.ship;
  }

  /** All prices, one per currency, sorted by currency code. */
  get prices(): Price[] {
    return [...this._prices.values()].sort((a, b) => a.currency.localeCompare(b.currency));
  }

  priceFor(currency: string): Price | undefined {
    return this._prices.get(currency);
  }

  /**
   * Resolves the current available-cabins count.
   *
   * - No live client wired (`'v1'`/`'v3'` formats): resolves immediately to
   *   the static {@link availableCabins} snapshot. Not cached — there is
   *   nothing to cache, it's a synchronous value wrapped in a Promise.
   * - Live client wired (`'swota'` format): invokes
   *   {@link ISwotaAvailabilityClient.getAvailableCabins} on the FIRST call
   *   and caches the resulting promise, so concurrent first-callers await the
   *   same in-flight request and later callers get the resolved result
   *   without re-invoking the client. The FIRST call also drives
   *   {@link availabilityState} through `'not-fetched'` → `'loading'` →
   *   `'loaded'`/`'failed'` (synchronously into `'loading'`, before any
   *   `await` point), notifying {@link onAvailabilityChange} listeners at
   *   each transition. Concurrent callers riding the same in-flight request
   *   never cause duplicate transitions or notifications. A `'failed'`
   *   offering is retryable, not terminal: a later call re-invokes the
   *   client rather than staying memoized as failed forever. Only `'loaded'`
   *   is truly final.
   *
   * This returned promise rejects if the live lookup fails (after this
   * class's own retry budget is exhausted) — callers are responsible for
   * handling that rejection. In particular, if you only need this call to
   * kick off the fetch (e.g. to start `'loading'` for a UI list) and don't
   * need the result here, you must still attach a `.catch()` when calling it
   * without `await`: a bare `void offering.getAvailableCabinsAsync()` with
   * no `.catch()` is an unhandled promise rejection, which crashes the
   * process under Node's default behavior. Always write it as:
   * ```ts
   * void offering.getAvailableCabinsAsync().catch(() => {});
   * ```
   * (see `utils/js/SDKCLI.js`'s `selectDeparture`, which does exactly this
   * when kicking off availability fetches for every offering in a
   * departure).
   */
  async getAvailableCabinsAsync(): Promise<number | null> {
    if (!this._liveClient) {
      return this.availableCabins;
    }
    // Entry guard checks `_availabilityState`, NOT `_liveAvailability` --
    // this is the reentrancy fix. `_transition('loading')` used to run
    // BEFORE `_liveAvailability` was assigned, which emitted 'change'
    // synchronously while `_liveAvailability` was still falsy; a listener
    // that reacted by calling this method again (a realistic UI-redraw
    // pattern) would see `!this._liveAvailability` still true, re-enter this
    // branch, transition to 'loading' again, emit again, and recurse without
    // bound -- reproduced as 1048 duplicate client calls and a stack
    // overflow.
    //
    // A 'not-fetched'/'failed' offering is what may start a fetch -- 'failed'
    // is retryable, not terminal, so a later call re-invokes the client
    // rather than staying memoized as failed forever; only 'loaded' is truly
    // final. The exact sequencing below is the safety-critical part:
    //   1. Flip `_availabilityState` to 'loading' as a raw synchronous field
    //      write -- no listener-invoking side effect yet.
    //   2. Assign `_liveAvailability` to the fetch's promise chain.
    //   3. ONLY THEN emit the 'loading' 'change' notification.
    // Steps 1 and 2 have nothing reentrant-triggering between them, so a
    // reentrant call arriving during step 3's emit (from inside a
    // synchronous listener) always observes `_availabilityState === 'loading'`
    // AND `_liveAvailability` already assigned -- it fails this guard and
    // falls straight through to `return this._liveAvailability` below,
    // riding the same in-flight fetch instead of starting a second one. No
    // code path can result in a second live fetch starting while one is
    // already logically in-flight, no matter what a synchronous 'change'
    // listener does.
    if (this._availabilityState === 'not-fetched' || this._availabilityState === 'failed') {
      this._availabilityState = 'loading';
      // A conforming `ISwotaAvailabilityClient` implementation returns a
      // Promise, but nothing stops one from throwing SYNCHRONOUSLY instead (a
      // plain, non-`async` function with a `throw` statement, executing
      // before any Promise is ever returned). Calling it directly here and
      // chaining `.then()`/`.catch()` off the return value would let that
      // throw propagate straight out of this method, uncaught -- and since
      // `_availabilityState` was already flipped to 'loading' above (which is
      // NOT one of the retryable entry states), the offering would be
      // permanently stuck 'loading' forever: no `_emitChange()`, no
      // subscriber notification, no way back in. The try/catch below
      // normalizes a synchronous throw into an ordinary rejected Promise, so
      // it flows through the same `.catch()` -> `_transition('failed')` path
      // as an async rejection, WITHOUT deferring the call itself into a
      // microtask (an alternative like `Promise.resolve().then(() => ...)`
      // would invoke the client one tick later than before, which is an
      // observable timing change -- this preserves synchronous invocation).
      // Mirrors .NET's guard for the same scenario.
      let clientPromise: Promise<number | null>;
      try {
        clientPromise = this._liveClient.getAvailableCabins(this.voyageId ?? '', this.code);
      } catch (err) {
        clientPromise = Promise.reject(err as unknown);
      }
      const promise = clientPromise
        .then((value) => {
          this._transition('loaded', value);
          return value;
        })
        .catch((err: unknown) => {
          this._transition('failed');
          throw err;
        });
      this._liveAvailability = promise;
      this._emitChange();
      return promise;
    }
    // Reachable only when `_availabilityState` is 'loading' or 'loaded' (the
    // only two states not handled above), both of which invariantly imply
    // `_liveAvailability` was already assigned by a previous call.
    return this._liveAvailability!;
  }

  /** Cabin description resolved for this offering's ship. */
  get description(): string[] {
    if (!this._cabinGrade) return [];
    const shipCode = this._departure.shipCode;
    return this._cabinGrade.descriptionsForShip(shipCode);
  }

  /** @internal */
  _setDeparture(departure: Departure): void {
    this._departure = departure;
  }

  /** @internal */
  _setCabinGrade(grade: CabinGrade): void {
    this._cabinGrade = grade;
  }

  /** @internal */
  _addPrice(currency: string, single: number | null, double: number | null): void {
    this._prices.set(currency, { currency, single, double });
  }
}

File: src/dotnet/ApiSdk/Data/CabinOffering.cs
using ApiSdk.Availability;

namespace ApiSdk.Data;

/// <summary>
/// Observable lifecycle of a <see cref="CabinOffering"/>'s availability figure.
/// See <see cref="CabinOffering.AvailabilityState"/>.
/// </summary>
public enum CabinAvailabilityState
{
    /// <summary>No live client configured (V1/V3 format): the static
    /// <see cref="CabinOffering.AvailableCabins"/> value is already final.
    /// Set at construction and never changes afterward.</summary>
    Static,

    /// <summary>Live client configured (SwOTA format) but
    /// <see cref="CabinOffering.GetAvailableCabinsAsync"/> hasn't been called
    /// yet. Initial state at construction for SwOTA offerings.</summary>
    NotFetched,

    /// <summary><see cref="CabinOffering.GetAvailableCabinsAsync"/> has been
    /// called and the fetch is in flight.</summary>
    Loading,

    /// <summary>The fetch completed successfully; the resolved value is
    /// available via <see cref="CabinOffering.LastKnownAvailableCabins"/>.</summary>
    Loaded,

    /// <summary>The fetch threw, after the client's own retry logic (if any)
    /// was exhausted.</summary>
    Failed,
}

/// <summary>
/// A cabin grade made available on a specific departure, with prices per
/// currency. The join between a Departure and a CabinGrade — the node where
/// pricing lives. Navigable to its departure, grade and ship.
/// </summary>
public sealed class CabinOffering
{
    private readonly Dictionary<string, Price> _prices = new();
    private readonly object _liveAvailabilityLock = new();
    private Departure _departure = null!;
    private CabinGrade? _cabinGrade;

    private readonly string? _voyageId;
    private readonly ISwOTAAvailabilityClient? _liveAvailabilityClient;
    private Task<int?>? _liveAvailabilityTask;

    // --- observable availability state --------------------------------------

    // Guarded by _liveAvailabilityLock — the same lock that already gates
    // "am I the first caller to start the fetch" below, so a state read never
    // observes a half-applied transition and the first-caller-wins race for
    // starting a fetch and the race for publishing its resulting state are
    // one and the same critical section.
    private CabinAvailabilityState _availabilityState;
    private int? _lastKnownAvailableCabins;

    /// <summary>
    /// Current point in the <see cref="CabinAvailabilityState"/> lifecycle.
    /// Thread-safe to read at any time; never blocks and never triggers a fetch.
    /// </summary>
    public CabinAvailabilityState AvailabilityState
    {
        get { lock (_liveAvailabilityLock) return _availabilityState; }
    }

    /// <summary>
    /// Synchronous, side-effect-free snapshot of whatever's currently known —
    /// safe to poll repeatedly (e.g. from a render loop). Never triggers a
    /// fetch. <c>null</c> while <see cref="CabinAvailabilityState.NotFetched"/>,
    /// <see cref="CabinAvailabilityState.Loading"/> or
    /// <see cref="CabinAvailabilityState.Failed"/>; the resolved value once
    /// <see cref="CabinAvailabilityState.Loaded"/>; the static
    /// <see cref="AvailableCabins"/> value immediately when
    /// <see cref="CabinAvailabilityState.Static"/>.
    /// </summary>
    public int? LastKnownAvailableCabins
    {
        get { lock (_liveAvailabilityLock) return _lastKnownAvailableCabins; }
    }

    /// <summary>
    /// Raised on the thread that completed the state transition (no
    /// thread-marshaling — this is a console TUI) whenever
    /// <see cref="AvailabilityState"/>/<see cref="LastKnownAvailableCabins"/>
    /// change: <c>NotFetched</c>→<c>Loading</c>, then
    /// <c>Loading</c>→<c>Loaded</c> or <c>Loading</c>→<c>Failed</c>. Never
    /// raised for <see cref="CabinAvailabilityState.Static"/> offerings.
    /// Invoked outside the internal lock, so a handler that calls back into
    /// this offering (e.g. to read <see cref="AvailabilityState"/>) can't deadlock.
    /// </summary>
    public event Action<CabinOffering>? AvailabilityChanged;

    /// <summary>Applies a transition and fires <see cref="AvailabilityChanged"/>.
    /// Caller must hold <see cref="_liveAvailabilityLock"/> to set the fields,
    /// but the event itself is raised after the lock is released (see call
    /// sites) so subscribers never run inside the lock.</summary>
    private void SetState(CabinAvailabilityState state, int? lastKnown)
    {
        _availabilityState = state;
        _lastKnownAvailableCabins = lastKnown;
    }

    /// <summary>Cabin grade code (source-market "Category", e.g. "DS").</summary>
    public string Code { get; }

    /// <summary>Human label (source-market "SuperCategory", e.g. "DARWIN SUITE").</summary>
    public string Name { get; }

    /// <summary>
    /// Static availability snapshot. Under V1 this is a real value from the
    /// flat file; under V3 there is no real field so the loader substitutes
    /// MaxOccupancy as a placeholder. Left untouched by the SwOTA live-lookup
    /// addition below — see <see cref="GetAvailableCabinsAsync"/> for the
    /// live counterpart.
    /// </summary>
    public int? AvailableCabins { get; }

    /// <param name="code">Cabin grade code — see <see cref="Code"/>.</param>
    /// <param name="name">Human label — see <see cref="Name"/>.</param>
    /// <param name="availableCabins">Static snapshot — see <see cref="AvailableCabins"/>.</param>
    /// <param name="voyageId">The departure/voyage identifier this offering is
    /// sold on, passed through to <paramref name="liveAvailabilityClient"/>
    /// lookups. Null/unused unless a live client is supplied.</param>
    /// <param name="liveAvailabilityClient">When non-null (only under
    /// <see cref="DataSourceFormat.SwOTA"/>), backs
    /// <see cref="GetAvailableCabinsAsync"/> with a live lookup instead of the
    /// static <paramref name="availableCabins"/> snapshot.</param>
    internal CabinOffering(
        string code,
        string name,
        int? availableCabins,
        string? voyageId = null,
        ISwOTAAvailabilityClient? liveAvailabilityClient = null)
    {
        Code = code;
        Name = name;
        AvailableCabins = availableCabins;
        _voyageId = voyageId;
        _liveAvailabilityClient = liveAvailabilityClient;

        // No lock needed here — this runs before the instance is published to
        // any other thread/caller.
        if (liveAvailabilityClient is null)
        {
            _availabilityState = CabinAvailabilityState.Static;
            _lastKnownAvailableCabins = availableCabins;
        }
        else
        {
            _availabilityState = CabinAvailabilityState.NotFetched;
            _lastKnownAvailableCabins = null;
        }
    }

    public Departure Departure => _departure;

    /// <summary>The cabin grade (null if the category is absent from cabingrades).</summary>
    public CabinGrade? CabinGrade => _cabinGrade;

    /// <summary>The ship this offering sails on, via its departure.</summary>
    public Ship? Ship => _departure.Ship;

    /// <summary>All prices, one per currency, ordered by currency code.</summary>
    public IReadOnlyList<Price> Prices =>
        _prices.Values.OrderBy(p => p.Currency, StringComparer.Ordinal).ToList();

    public Price? PriceFor(string currency) =>
        _prices.TryGetValue(currency, out var price) ? price : null;

    /// <summary>Cabin description resolved for this offering's ship.</summary>
    public IReadOnlyList<string> Description =>
        _cabinGrade?.DescriptionsForShip(_departure.ShipCode) ?? Array.Empty<string>();

    internal void SetDeparture(Departure departure) => _departure = departure;

    internal void SetCabinGrade(CabinGrade grade) => _cabinGrade = grade;

    internal void AddPrice(string currency, double? single, double? @double) =>
        _prices[currency] = new Price(currency, single, @double);

    /// <summary>
    /// Live-or-static cabin availability.
    /// <list type="bullet">
    /// <item>No live client configured (V1/V3 formats): returns the static
    /// <see cref="AvailableCabins"/> value immediately, no caching needed.
    /// <see cref="AvailabilityState"/> is <see cref="CabinAvailabilityState.Static"/>
    /// already and this call is a no-op with respect to it.</item>
    /// <item>Live client configured (SwOTA format): the first call invokes the
    /// client and caches the resulting task, so concurrent first-callers await
    /// the same in-flight call; later calls return the cached, already-completed
    /// result without re-invoking the client. That same first caller —
    /// determined atomically under <see cref="_liveAvailabilityLock"/>, so
    /// concurrent racers never both "win" — synchronously flips
    /// <see cref="AvailabilityState"/> from <see cref="CabinAvailabilityState.NotFetched"/>
    /// to <see cref="CabinAvailabilityState.Loading"/> and fires
    /// <see cref="AvailabilityChanged"/> once for that transition BEFORE the
    /// live client is ever invoked (see <see cref="StartFetch"/>) — so this is
    /// true unconditionally, whether the client resolves normally, returns a
    /// task that later faults, or throws synchronously the instant it's
    /// called; there is no path by which the terminal transition can run
    /// ahead of, or collapse, the Loading one. When the underlying call
    /// resolves or throws, the state moves once more to
    /// <see cref="CabinAvailabilityState.Loaded"/> or
    /// <see cref="CabinAvailabilityState.Failed"/>, and the
    /// <see cref="Task{TResult}"/> returned/memoized here completes FIRST — so
    /// every caller (the first one and every concurrent or later one awaiting
    /// the same memoized task) is guaranteed to observe the terminal state
    /// already applied by the time its own await returns; there is no window
    /// where a caller's result is available but
    /// <see cref="AvailabilityState"/> still reads <c>Loading</c>. Only *after*
    /// that task has completed does <see cref="AvailabilityChanged"/> fire for
    /// that transition (its second and final firing) — this ordering is what
    /// lets a subscriber that synchronously blocks on this same task from
    /// within its own handler do so without deadlocking, since the task is
    /// already complete by the time any handler runs. Concurrent callers that
    /// lose the race never drive the state machine themselves — they just
    /// await the same task and observe its one set of transitions.</item>
    /// <item>A <see cref="CabinAvailabilityState.Failed"/> offering is
    /// retryable, not terminal: a later call re-invokes the client rather
    /// than staying memoized as failed forever. Only
    /// <see cref="CabinAvailabilityState.Loaded"/> is truly final.</item>
    /// </list>
    /// </summary>
    /// <param name="ct">Accepted for API consistency/future use, but not
    /// currently wired to anything: it is not observed anywhere in this
    /// method's body, so it does not cancel even the caller's own await of
    /// the returned task, let alone the underlying SWOTA fetch. The fetch is
    /// shared/memoized across every concurrent or later caller, so honoring
    /// any one caller's token on the shared client call would let that
    /// caller's cancellation poison the result for every other caller riding
    /// the same memoized task; the actual client call always runs with
    /// <see cref="CancellationToken.None"/> regardless of what's passed
    /// here.</param>
    public Task<int?> GetAvailableCabinsAsync(CancellationToken ct = default)
    {
        if (_liveAvailabilityClient is null) return Task.FromResult(AvailableCabins);

        Task<int?> task;
        var isFirstCaller = false;
        TaskCompletionSource<int?>? tcs = null;
        lock (_liveAvailabilityLock)
        {
            // A Failed offering is retryable -- only Loaded is truly terminal.
            // Without this, a transient SWOTA failure would memoize "Failed"
            // forever with no way for a later call to ever try again.
            if (_liveAvailabilityTask is null || _availabilityState == CabinAvailabilityState.Failed)
            {
                isFirstCaller = true;
                SetState(CabinAvailabilityState.Loading, null);

                // Memoize a TaskCompletionSource's Task, NOT the live client's
                // task -- the client hasn't even been invoked yet at this
                // point. This is what lets the Loading transition be fully
                // applied AND announced (via RaiseAvailabilityChanged, below,
                // once the lock is released) strictly BEFORE the client is
                // ever called (see StartFetch). Without this indirection, a
                // client that throws synchronously would let the terminal
                // Loaded/Failed transition run to completion (including firing
                // AvailabilityChanged) before this method ever releases the
                // lock or announces Loading -- reentering the lock (harmless
                // on its own, since Monitor is same-thread-reentrant) but
                // firing the terminal event BEFORE the Loading event, and
                // collapsing Loading to a state no external caller could ever
                // observe. RunContinuationsAsynchronously keeps any awaiter's
                // continuation off of whatever thread completes the fetch,
                // matching the "no thread-marshaling" behaviour the rest of
                // this type already documents for AvailabilityChanged.
                tcs = new TaskCompletionSource<int?>(TaskCreationOptions.RunContinuationsAsynchronously);
                _liveAvailabilityTask = tcs.Task;
            }
            task = _liveAvailabilityTask;
        }

        // Fired outside the lock: AvailabilityChanged is arbitrary subscriber
        // code (e.g. a TUI redraw) and must never run while holding a lock
        // other callers may be waiting on. Only after this has fully returned
        // (including swallowing any subscriber exception, see
        // RaiseAvailabilityChanged) do we ever invoke the live client -- see
        // StartFetch -- so the Loading announcement can never be raced or
        // skipped by a synchronous throw.
        if (isFirstCaller)
        {
            RaiseAvailabilityChanged();
            StartFetch(tcs!);
        }

        return task;
    }

    /// <summary>Invokes the live client -- catching a synchronous throw
    /// exactly like an asynchronous fault, so both paths drive the same
    /// Loading -> Loaded/Failed transition -- and arranges for the result to
    /// complete <paramref name="tcs"/>. Only ever called once per offering per
    /// fetch attempt, by the single first caller identified in
    /// <see cref="GetAvailableCabinsAsync"/>, and only after that caller has
    /// already released <see cref="_liveAvailabilityLock"/> and announced the
    /// Loading transition -- see the doc comment there for why that ordering
    /// matters.</summary>
    private void StartFetch(TaskCompletionSource<int?> tcs)
    {
        Task<int?> started;
        try
        {
            // CancellationToken.None, deliberately not any caller's ct — see
            // the doc comment on GetAvailableCabinsAsync.
            started = _liveAvailabilityClient!.GetAvailableCabinsAsync(_voyageId ?? string.Empty, Code, CancellationToken.None);
        }
        catch (Exception ex)
        {
            // A client that throws synchronously instead of returning a
            // faulted task still needs to drive the same Loading -> Failed
            // transition below.
            started = Task.FromException<int?>(ex);
        }

        // Deliberately not awaited: this method must return immediately
        // (nothing here needs to observe the fetch's outcome), and the
        // continuation below never lets an exception escape it -- every path
        // through it either returns normally or is caught -- so there is
        // nothing for an UnobservedTaskException to complain about.
        _ = ApplyTerminalTransition(started, tcs);
    }

    /// <summary>Awaits the underlying client call, applies the Loaded/Failed
    /// transition, completes <paramref name="tcs"/> (and therefore every
    /// caller's memoized task — see <see cref="GetAvailableCabinsAsync"/>)
    /// with the same result (or the same exception) as
    /// <paramref name="inner"/>, then fires
    /// <see cref="AvailabilityChanged"/>.
    ///
    /// The state transition is fully applied (including which terminal state
    /// was reached) BEFORE <paramref name="tcs"/> is completed, and
    /// <paramref name="tcs"/> is completed BEFORE
    /// <see cref="AvailabilityChanged"/> is raised -- so every caller (the
    /// first one and every concurrent or later one awaiting the same
    /// memoized task) is guaranteed to observe the terminal state already
    /// applied by the time its own await returns. This ordering also matters
    /// for a subscriber that itself synchronously blocks on
    /// <paramref name="tcs"/>'s task from within its own
    /// <see cref="AvailabilityChanged"/> handler (e.g. by calling
    /// <c>.Result</c>/<c>.Wait()</c> or otherwise re-entering
    /// <see cref="GetAvailableCabinsAsync"/> synchronously): because the task
    /// is already complete by the time the event fires, that handler
    /// observes an already-finished task instead of blocking on one that can
    /// only be completed by a continuation still waiting to run, which is
    /// what would deadlock. What prevents a subscriber's exception from
    /// being mis-attributed to the fetch and flipping an already-successful
    /// Loaded result back to Failed is that
    /// <see cref="RaiseAvailabilityChanged"/> swallows subscriber exceptions
    /// itself before they ever reach this method's catch block, so they can't
    /// propagate out of this method either.</summary>
    private async Task ApplyTerminalTransition(Task<int?> inner, TaskCompletionSource<int?> tcs)
    {
        try
        {
            var result = await inner.ConfigureAwait(false);
            lock (_liveAvailabilityLock)
            {
                SetState(CabinAvailabilityState.Loaded, result);
            }
            tcs.TrySetResult(result);
            RaiseAvailabilityChanged();
        }
        catch (Exception ex)
        {
            lock (_liveAvailabilityLock)
            {
                SetState(CabinAvailabilityState.Failed, null);
            }
            tcs.TrySetException(ex);
            RaiseAvailabilityChanged();
        }
    }

    /// <summary>Raises <see cref="AvailabilityChanged"/>, catching and
    /// discarding any exception a subscriber throws so it can never
    /// propagate back into fetch-state logic (see <see cref="ApplyTerminalTransition"/>,
    /// which calls this only after the Loaded/Failed transition is already
    /// fully applied) or out to a caller merely awaiting the availability
    /// result.</summary>
    private void RaiseAvailabilityChanged()
    {
        try
        {
            AvailabilityChanged?.Invoke(this);
        }
        catch
        {
            // Subscriber misbehaved (e.g. a TUI redraw handler threw). The
            // fetch itself already succeeded or failed on its own merits by
            // the time this runs -- a subscriber exception must not corrupt
            // that outcome, so it's swallowed here rather than surfaced.
        }
    }
}
